zerver/lib/push_notifications.py

```diff
@@ -1188,3 +1188,55 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
     user_identity = UserPushIdentityCompat(user_id=user_profile.id)
     send_apple_push_notification(user_identity, apple_devices, apns_payload)
     send_android_push_notification(user_identity, android_devices, gcm_payload, gcm_options)
+
+
+def send_test_push_notification_directly_to_device(
+    user_identity: UserPushIdentityCompat,
+    device: DeviceToken,
+    base_payload: Dict[str, Any],
+    remote: Optional["RemoteZulipServer"] = None,
+) -> None:
+    payload = base_payload
+    payload["event"] = "test-by-device-token"
+
+    if device.kind == RemotePushDeviceToken.APNS:
+        realm_uri = base_payload["realm_uri"]
+        apns_data = {
+            "alert": {
+                "title": _("Test notification"),
+                "body": _("This is a test notification from {realm_uri}.").format(
+                    realm_uri=realm_uri
+                ),
+            },
+            "sound": "default",
+            "custom": {"zulip": payload},
+        }
+        send_apple_push_notification(user_identity, [device], apns_data, remote=remote)
+    else:
+        assert device.kind == RemotePushDeviceToken.GCM
+        payload["time"] = datetime_to_timestamp(timezone_now())
```

Comment:
I don't think this is up to date with fd6091ad17b5e9e9e786418e3e8053332c67c0f9 and the code in general.  `get_base_payload` (passed in as `base_payload` here) is going to contain a `time` key already, with the appropriate granularity.

Reply:
Hmm, `time` is inserted in `get_message_payload` currently, not `get_base_payload`. But possibly we should move that into `get_base_payload`? That'll also mean `time` will be added in the `get_remove_payload_apns` `get_remove_payload_gcm` codepaths, but probably that's fine and shouldn't cause compatibility issues with the apps?

Reply:
Ah, you're right that I was confusing `get_message_payload` and `get_base_payload`.

As for if it'd be a problem for `remove` payloads, that [sounds like an #api design question](https://github.com/zulip/zulip/pull/26768#issuecomment-1741313824).